rust_src/src/eval.rs

```diff
@@ -1225,4 +1225,29 @@ pub fn catch(args: LispCons) -> LispObject {
     unsafe { internal_catch(val, Some(Fprogn), body) }
 }
 
+/// Signal an error.  Args are ERROR-SYMBOL and associated DATA. This
+/// function does not return.
+///
+/// An error symbol is a symbol with an `error-conditions' property
+/// that is a list of condition names.  A handler for any of those
+/// names will get to handle this signal.  The symbol `error' should
+/// normally be one of them.
+///
+/// DATA should be a list.  Its elements are printed as part of the
+/// error message.  See Info anchor `(elisp)Definition of signal' for
+/// some details on how this error message is constructed.
+/// If the signal is handled, DATA is made available to the handler.
+/// See also the function `condition-case'.
+#[lisp_fn]
+pub fn signal(error_symbol: LispObject, data: LispObject) -> ! {
+    xsignal(error_symbol, data)
```

Comment:
I like the idea of standardizing on one implementation. However, `xsignal` is not it. In Rust we already have a `xisgnal` macro which handles arbitrary arguments. Now the developer needs to switch between a macro and a function of the same name. That is not good practice.

By porting `Fsignal` you also created `signal`. Which means there is no need for an internal function. Let `signal` be the implementation and `xsignal!` be the macro we prefer everyone uses to access it. When we port C code we keep the `xsignal` calls and adjust them to be `xsignal!` macro calls. So leave the `xsignal` function in place for C.

Reply:
Thanks, that makes sense!
I had some issues because the Rust-generated `Fsignal` is not marked noreturn in the type system.
This is not trivial to change because the signatures in `struct Lisp_Subr` have to return a `Lisp_Object` (It doesn't seem possible to make the members noreturn).

Anyway, I have cleaned up the mess now, and I just use `std::unreachable()!` after calling `Fsignal`.
This way there are far fewer changes and new code will use `xsignal!` anyway (so people won't need to add `std::unreachable!()`)